query/src/org/labkey/query/controllers/QueryController.java

```diff
@@ -8216,6 +8168,294 @@ public Object execute(ParseForm form, BindException errors) throws Exception
+
+                if (file instanceof MultipartFile || file instanceof SpringAttachmentFile)
+                {
+                    String fileName;
+                    if (file instanceof MultipartFile f)
+                        fileName = f.getName();
+                    else
+                    {
+                        SpringAttachmentFile f = (SpringAttachmentFile) file;
+                        fileName = f.getFilename();
+                    }
+                    String fileNameValidation = FileUtil.validateFileName(fileName);
+                    if (!StringUtils.isEmpty(fileNameValidation))
+                        throw new IllegalArgumentException(fileNameValidation);
+
+                    FileLike uploadDir = ensureUploadDirectory(container, kind.getDomainFileDirectory());
+                    uploadDir = uploadDir.resolveChild("_templates");
+                    Object savedFile = saveFile(user, container, "template file", file, uploadDir);
+                    Path savedFilePath;
+
+                    if (savedFile instanceof File ioFile)
+                        savedFilePath = ioFile.toPath();
+                    else if (savedFile instanceof FileLike fl)
+                        savedFilePath = fl.toNioPathForRead();
+                    else
+                        throw UnexpectedException.wrap(null,"Unable to upload template file.");
+
+                    templateUrl = fcs.getWebDavUrl(savedFilePath, container, FileContentService.PathType.serverRelative).toString();
+                }
+
+                uploadedTemplates.add(Pair.of(templateLabel, templateUrl));
+            }
+            return uploadedTemplates;
+        }
+
+        @Override
+        public Object execute(QueryImportTemplateForm form, BindException errors) throws ValidationException, QueryUpdateServiceException, SQLException, ExperimentException, MetadataUnavailableException
+        {
+            User user = getUser();
+            Container container = getContainer();
+            String schemaName = form.getSchemaName();
+            String queryName = form.getQueryName();
+
+            List<Pair<String, String>> updatedTemplates = getUploadedTemplates(form, _kind);
+
+            List<Pair<String, String>> existingTemplates = _tInfo.getImportTemplates(getViewContext());
+            List<Pair<String, String>> existingCustomTemplates = new ArrayList<>();
+            for (Pair<String, String> template_ : existingTemplates)
+            {
+                if (!template_.second.toLowerCase().contains("exportexceltemplate"))
+                    existingCustomTemplates.add(template_);
+            }
+            if (!updatedTemplates.equals(existingCustomTemplates))
+            {
+                QueryDef queryDef = null;
+                if (queryDef == null)
+                    queryDef = QueryManager.get().getQueryDef(container, schemaName, queryName, false);

```

Comment:
queryDef is always null here so not sure that if statement is needed. 
```suggestion
                QueryDef queryDef = QueryManager.get().getQueryDef(container, schemaName, queryName, false);

```